                                                     MISSING VALUE HANDLING STRATEGY
                                                     -------------------------------
In this notebook, missing values identified in the previous analysis are used and we observe and select specific strategies that can be used in                                               those missing values using predefined strategies.

The goal is to clean the dataset while preserving as much information as possible, without performing any encoding, scaling, or model-related steps.

In [1]:
import pandas as pd

columns = [
    "age", "workclass", "fnlwgt", "education", "education_num",
    "marital_status", "occupation", "relationship", "race", "sex",
    "capital_gain", "capital_loss", "hours_per_week", "native_country",
    "income"
]

df = pd.read_csv(
    "adult_income_raw.csv",
    sep=", ",
    engine="python",
    header=None,
    names=columns,
    na_values="?"
)
print("the dataset is loaded successfully ")

the dataset is loaded successfully 


In [2]:
df.isnull().sum()

age                  0
workclass         1837
fnlwgt               1
education            1
education_num        1
marital_status       1
occupation        1844
relationship         1
race                 1
sex                  1
capital_gain         1
capital_loss         1
hours_per_week       1
native_country     584
income               1
dtype: int64

Missing Value Reconfirmation:
The missing value distribution matches the observations made during the missing value analysis phase. Both numerical and categorical columns contain missing values.

In [3]:
df.dtypes

age                object
workclass          object
fnlwgt            float64
education          object
education_num     float64
marital_status     object
occupation         object
relationship       object
race               object
sex                object
capital_gain      float64
capital_loss      float64
hours_per_week    float64
native_country     object
income             object
dtype: object

In [4]:
numerical_cols= df.select_dtypes(include=['int64' , 'float64'] ).columns
categorical_cols=df.select_dtypes(include='object').columns

Now at this step the numerical columns and categorical columns are separated for finding further of how many missing values does numerical columns have and categrocial columns have 

In [5]:
numerical_missing_cols = df[numerical_cols].columns[
    df[numerical_cols].isnull().sum() > 0
]

In [6]:
numerical_missing_cols

Index(['fnlwgt', 'education_num', 'capital_gain', 'capital_loss',
       'hours_per_week'],
      dtype='object')

In [7]:
categorical_missing_cols= df[categorical_cols].columns[df[categorical_cols].isnull().sum()>0]

In [8]:
numerical_cols

Index(['fnlwgt', 'education_num', 'capital_gain', 'capital_loss',
       'hours_per_week'],
      dtype='object')

In [9]:
missing_num_data = df[numerical_missing_cols]

In [10]:
missing_num_data.head(30)

,fnlwgt,education_num,capital_gain,capital_loss,hours_per_week
0,77516.0,13.0,2174.0,0.0,40.0
1,83311.0,13.0,0.0,0.0,13.0
2,215646.0,9.0,0.0,0.0,40.0
3,234721.0,7.0,0.0,0.0,40.0
4,338409.0,13.0,0.0,0.0,40.0
5,284582.0,14.0,0.0,0.0,40.0
6,160187.0,5.0,0.0,0.0,16.0
7,209642.0,9.0,0.0,0.0,45.0
8,45781.0,14.0,14084.0,0.0,50.0
9,159449.0,13.0,5178.0,0.0,40.0


In [11]:
missing_num_data["fnlwgt"].median()

178356.0

In [12]:
missing_num_data["fnlwgt"].mean()

np.float64(189778.36651208502)

We can check if the column is symmetric or clean by using this method also that if there is a large difference between the mean and median of that particular column

In [13]:
missing_num_data["education_num"].median()

10.0

In [14]:
missing_num_data["education_num"].mean()

np.float64(10.0806793403151)

Here i can see both mean and median are equal for this column here mean can be applied but still median also can be applied as it has no benefit by applying mean we should chase benefits of everything here even after using mean it stays the same until theres no strong reason we will not be using mean in our columns to maintain consistency we should use median here

In [15]:
missing_num_data["capital_gain"].median()

0.0

In [16]:
missing_num_data["capital_gain"].mean()

np.float64(1077.6488437087312)

Median is applicable

In [17]:
missing_num_data["capital_loss"].median()

0.0

In [18]:
missing_num_data["capital_loss"].mean()

np.float64(87.303829734959)

Median is applicable

In [19]:
missing_num_data["hours_per_week"].median()

40.0

In [20]:
missing_num_data["hours_per_week"].mean()

np.float64(40.437455852092995)

Median still is applicable and best solution

In [21]:
missing_cat_data = df[categorical_missing_cols]

In [22]:
missing_cat_data.head(30)

,workclass,education,marital_status,occupation,relationship,race,sex,native_country,income
0,State-gov,Bachelors,Never-married,Adm-clerical,Not-in-family,White,Male,United-States,"<=50K"""
1,Self-emp-not-inc,Bachelors,Married-civ-spouse,Exec-managerial,Husband,White,Male,United-States,"<=50K"""
2,Private,HS-grad,Divorced,Handlers-cleaners,Not-in-family,White,Male,United-States,"<=50K"""
3,Private,11th,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,United-States,"<=50K"""
4,Private,Bachelors,Married-civ-spouse,Prof-specialty,Wife,Black,Female,Cuba,"<=50K"""
5,Private,Masters,Married-civ-spouse,Exec-managerial,Wife,White,Female,United-States,"<=50K"""
6,Private,9th,Married-spouse-absent,Other-service,Not-in-family,Black,Female,Jamaica,"<=50K"""
7,Self-emp-not-inc,HS-grad,Married-civ-spouse,Exec-managerial,Husband,White,Male,United-States,">50K"""
8,Private,Masters,Never-married,Prof-specialty,Not-in-family,White,Female,United-States,">50K"""
9,Private,Bachelors,Married-civ-spouse,Exec-managerial,Husband,White,Male,United-States,">50K"""


In [23]:
missing_cat_data.isnull().sum()

workclass         1837
education            1
marital_status       1
occupation        1844
relationship         1
race                 1
sex                  1
native_country     584
income               1
dtype: int64

In [24]:
missing_cat_data["workclass"]

0               State-gov
1        Self-emp-not-inc
2                 Private
3                 Private
4                 Private
               ...       
32557             Private
32558             Private
32559             Private
32560        Self-emp-inc
32561                None
Name: workclass, Length: 32562, dtype: object

As we can observe there are total 1837 missing values in the column workclass and the total number of rows is 32562 rows the total percentage of missing values is <5% so we can use MODE here

In [25]:
missing_cat_data["education"]

0        Bachelors
1        Bachelors
2          HS-grad
3             11th
4        Bachelors
           ...    
32557      HS-grad
32558      HS-grad
32559      HS-grad
32560      HS-grad
32561         None
Name: education, Length: 32562, dtype: object

Same amount of rows are there but missing valeus are almost negligible so we can fill it as MISSING or UNKNOWN

In [26]:
missing_cat_data["marital_status"]

0             Never-married
1        Married-civ-spouse
2                  Divorced
3        Married-civ-spouse
4        Married-civ-spouse
                ...        
32557    Married-civ-spouse
32558               Widowed
32559         Never-married
32560    Married-civ-spouse
32561                  None
Name: marital_status, Length: 32562, dtype: object

Same filling as MISSING or UNKNOWN is applicable

In [27]:
missing_cat_data["occupation"]

0             Adm-clerical
1          Exec-managerial
2        Handlers-cleaners
3        Handlers-cleaners
4           Prof-specialty
               ...        
32557    Machine-op-inspct
32558         Adm-clerical
32559         Adm-clerical
32560      Exec-managerial
32561                 None
Name: occupation, Length: 32562, dtype: object

Here MODE is applicable as it has <5% missing values

In [28]:
missing_cat_data["relationship"]

0        Not-in-family
1              Husband
2        Not-in-family
3              Husband
4                 Wife
             ...      
32557          Husband
32558        Unmarried
32559        Own-child
32560             Wife
32561             None
Name: relationship, Length: 32562, dtype: object

Here we can fill MISSING or UNKNOWN

In [29]:
missing_cat_data["race"]

0        White
1        White
2        White
3        Black
4        Black
         ...  
32557    White
32558    White
32559    White
32560    White
32561     None
Name: race, Length: 32562, dtype: object

Here we can fill MISSING or UNKNOWN

In [30]:
missing_cat_data["sex"]

0          Male
1          Male
2          Male
3          Male
4        Female
          ...  
32557      Male
32558    Female
32559      Male
32560    Female
32561      None
Name: sex, Length: 32562, dtype: object

Here we can fill MISSING or UNKNOWN

In [31]:
missing_cat_data["native_country"].value_counts()

native_country
United-States                 29170
Mexico                          643
Philippines                     198
Germany                         137
Canada                          121
Puerto-Rico                     114
El-Salvador                     106
India                           100
Cuba                             95
England                          90
Jamaica                          81
South                            80
China                            75
Italy                            73
Dominican-Republic               70
Vietnam                          67
Guatemala                        64
Japan                            62
Poland                           60
Columbia                         59
Taiwan                           51
Haiti                            44
Iran                             43
Portugal                         37
Nicaragua                        34
Peru                             31
France                           29
Greece       

There was 1% missing values in this column according to its size but we can use mode here as there is a dominating categorie United_States

In [32]:
missing_cat_data["income"]

0        <=50K"
1        <=50K"
2        <=50K"
3        <=50K"
4        <=50K"
          ...  
32557     >50K"
32558    <=50K"
32559    <=50K"
32560     >50K"
32561      None
Name: income, Length: 32562, dtype: object

Here we can fill MISSING or UNKNOWN

Final Missing Value Handling Strategy

After analyzing the distribution, skewness, and missing percentage of each feature, the following imputation strategies have been finalized:

🔹 Numerical Features

Skewed distributions → Median imputation

Approximately symmetric distributions → Mean imputation

Zero-inflated features (majority values = 0) → Median imputation

Justification:
Median is robust to outliers and skewed data, while mean is appropriate only when the distribution is reasonably symmetric and not affected by extreme values.

🔹 Categorical Features

Low to moderate missing percentage → Most Frequent (Mode) imputation

Features where missing may represent an implicit category → Constant value imputation ("Unknown")

Justification:
Mode imputation preserves the dominant category distribution.
Constant imputation prevents distortion when missingness may carry structural meaning.

🔹 Implementation Plan

These strategies will be implemented in the final preprocessing engine using:

SimpleImputer(strategy="mean")

SimpleImputer(strategy="median")

SimpleImputer(strategy="most_frequent")

SimpleImputer(strategy="constant", fill_value="Unknown")

All imputations will be applied through a structured Pipeline and ColumnTransformer to ensure consistency, automation, and reusability.